sdk/core/core/src/az_log.c

```diff
@@ -10,62 +10,71 @@
 #include <az_log_internal.h>
 #include <az_span.h>
 
-#include <stdbool.h>
 #include <stddef.h>
 
 #include <_az_cfg.h>
 
 static az_log_classification const* _az_log_classifications = NULL;
-static size_t _az_log_classifications_length = 0;
-static az_log_fn _az_log_listener = NULL;
+static az_log_message_fn _az_log_message_callback = NULL;
 
-void az_log_set_classifications(
-    az_log_classification const* classifications,
-    size_t classifications_length)
+void az_log_set_classifications(az_log_classification const classifications[])
 {
-  // TODO: thread safety
   _az_log_classifications = classifications;
-  _az_log_classifications_length = classifications_length;
 }
 
-void az_log_set_listener(az_log_fn listener)
+void az_log_set_callback(az_log_message_fn az_log_message_callback)
 {
-  // TODO: thread safety
-  _az_log_listener = listener;
+  _az_log_message_callback = az_log_message_callback;
 }
 
 void az_log_write(az_log_classification classification, az_span message)
 {
-  // TODO: thread safety
-  if (_az_log_listener != NULL && az_log_should_write(classification))
+  if (_az_log_message_callback != NULL && az_log_should_write(classification))
   {
-    (*_az_log_listener)(classification, message);
+    int32_t const length = az_span_length(message);
+    int32_t const capacity = az_span_capacity(message);
+    // Do nothing if span is empty or if truncating it with zero would result in empty span.
+    if (length > 0 && capacity > 1)
+    {
+      int32_t const z_pos = length < capacity ? length + 1 : capacity - 1;
+      char* const buf = (char*)az_span_ptr(message);
+
+      // Don't write 0 if it's already there, for instance when the code logs a string literal span,
+      // don't try to modify it as there is a chance it resides in .text.
+      if (buf[z_pos] != 0)
+      {
+        buf[z_pos] = 0;
+      }
+
+      _az_log_message_callback(classification, buf, z_pos - 1);
```

Comment:
I don't think you want -1 here. The z_pos is the length.

Reply:
No longer applies.